crates/dbsp/src/trace/mod.rs

```diff
@@ -692,39 +738,38 @@ where
 }
 
-/// Copies a batch.  This uses the [`Builder`] interface to produce the output
-/// batch, which only supports producing batches with a single fixed timestamp,
-/// which must be supplied as `timestamp`.  To avoid discarding meaningful
-/// timestamps in the input, the input batch's time type must be `()`.
-///
-/// This is useful for adding a timestamp to a batch, or for converting between
-/// different batch implementations (e.g. writing an in-memory batch to disk).
+/// Merges all of the batches in `batches` and returns the merged result.
 ///
-/// TODO: for adding a timestamp to a batch, this could be implemented more
-/// efficiently by having a special batch type where all updates have the same
-/// timestamp, as this is the only kind of batch that we ever create directly in
-/// DBSP; batches with multiple timestamps are only created as a result of
-/// merging.  The main complication is that we will need to extend the trace
-/// implementation to work with batches of multiple types.  This shouldn't be
-/// too hard and is on the todo list.
-pub fn copy_batch<BI, TS, BO>(batch: &BI, timestamp: &TS, factories: &BO::Factories) -> BO
+/// This implements a multiway merge rather than iterating a 2-way merge.  This
+/// is theoretically and asymptotically more efficient, but it is practically
+/// much more efficient for the case where cloning items in the batch is
+/// expensive, because 2-way merges will clone each data item `lg n` times
+/// whereas an N-way merge will only clone them once each. For a 16-way merge,
+/// that's a 4x reduction.
+pub fn merge_untimed_batches<B, T>(factories: &B::Factories, batches: T) -> B
 where
-    TS: Timestamp,
-    BI: BatchReader<Time = ()>,
-    BO: Batch<Key = BI::Key, Val = BI::Val, Time = TS, R = BI::R>,
+    T: IntoIterator<Item = B>,
+    B: Batch<Time = ()>,
 {
-    let mut builder = BO::Builder::with_capacity(factories, timestamp.clone(), batch.len());
-    let mut cursor = batch.cursor();
-    let mut weight = batch.factories().weight_factory().default_box();
-    while cursor.key_valid() {
-        while cursor.val_valid() {
-            cursor.weight().clone_to(&mut weight);
-            builder.push_refs(cursor.key(), cursor.val(), &weight);
-            cursor.step_val();
+    let mut batches: Vec<_> = batches.into_iter().filter(|b| !b.is_empty()).collect();
+    match batches.len() {
+        0 => B::dyn_empty(factories),
+        1 => batches.pop().unwrap(),
+        2 => {
+            // Presumably the specialized merge implementation for the batch
+            // type can do better than our general algorithm.
+            merge_batches(factories, batches)
         }
-        cursor.step_key();
+        _ => B::from_cursor(
+            CursorList::new(
+                factories.weight_factory(),
+                batches.iter().map(|b| b.cursor()).collect(),
+            ),
+            &(),
+            factories,
+            batches.iter().map(|b| b.len()).sum(),
```

Comment:
If we use this implementation in the merger in the future, we may want to be more careful about over-allocating here.